Completed all steps of module 4 project exercise (except from Stretch Goal: Provision as root)

gmoraita opened the pull request.

gmoraita committed:
module 4 - all steps completed

hugh-emerson commented on README.md:
```diff
@@ -70,3 +70,32 @@ You should see output similar to the following:
+* Download and install Vagrant: https://www.vagrantup.com/downloads 
+* Open powershell 
+* Go to the root of your TODO app
+* Run the following on your command line to create the hypervisor with the TODO app installed and running:
```
Just a correction with the terminology - you are creating a VM, not a hypervisor. 

A hypervisor is the software that handles the virtualisation/runs the VMs.

gmoraita replied:
Sorry yes - agreed.

hugh-emerson commented on Vagrantfile:
```diff
@@ -0,0 +1,118 @@
+  # boxes at https://vagrantcloud.com/search.
+  config.vm.box = "hashicorp/bionic64"
+
+  # Disable automatic box update checking. If you disable this, then
+  # boxes will only be checked for updates when the user runs
+  # `vagrant box outdated`. This is not recommended.
+  # config.vm.box_check_update = false
+
```
I would recommend removing all of these comments. 
Your codebase isn't really the best place to store Vagrant document once you're done creating the Vagrantfile. It just obscures your actual config.

At most you'd keep the comment with a link to the actual docs.

gmoraita replied:
Done

hugh-emerson commented on Vagrantfile:
```diff
@@ -0,0 +1,118 @@
+	# Install pyenv
+	git clone https://github.com/pyenv/pyenv.git ~/.pyenv
+	
+	# Set the profile and the env if this is not done before
+	if [ ! $PYENV_ROOT ]
```
Nice one!
You could also have moved the git clone into the if block to avoid seeing a red line in the vagrant output, and feels a bit nicer.

gmoraita replied:
Done

hugh-emerson commented on Vagrantfile:
```diff
@@ -0,0 +1,118 @@
+	
+	# Install gunicorn
+	cd /vagrant
+	poetry run pip install gunicorn
```
This means the version of gunicorn isn't source controlled.

I don't think there's any significant cost to just adding gunicorn to the pyproject.toml like any other dependency.

gmoraita replied:
Done

hugh-emerson commented on README.md:
````diff
@@ -70,3 +70,32 @@ You should see output similar to the following:
+```
+* Once complete login to the hypervisor created (username & password: vagrant/vagrant)
+```bash
+vagrant ssh
+```
+* Then tail the gunicorn access logs:
+```bash
+cd /vagrant
+tail -f ./gunicorn-access.log
````
You can tail it from your own machine rather than the VM since you're creating the log file in the shared directory. Though the command would be different in powershell.

hugh-emerson commented on Vagrantfile:
```diff
@@ -0,0 +1,118 @@
+	trigger.name = "Launching App"
+	trigger.info = "Running the TODO app setup script"
+	trigger.run_remote = {privileged: false, inline: "
+	# Install dependencies and launch
+	cd /vagrant
+	poetry install
+
+	# uncomment the below if you want to run flask instead of gunicorn
+	# poetry run flask run --host 0.0.0.0
+	
+	# Run gunicorn
+	poetry run gunicorn -b 0.0.0.0:5000 'todo_app.app:app' --daemon --access-logfile gunicorn-access.log --log-file gunicorn.log
```
This section should be indented another level.

gmoraita replied:
Done

hugh-emerson commented on .gitignore:
```diff
@@ -118,3 +118,12 @@ env/
+.vagrant
+
+### Poetry ###
+poetry.lock
```
Actually the lock file shouldn't be gitignored. It lets you control the versions of dependencies precisely, to avoid any surprising differences between servers.

gmoraita replied:
Fixed

hugh-emerson commented on README.md:
```diff
@@ -70,3 +70,32 @@ You should see output similar to the following:
+You can create a new Hypervisor installed with your app and in running mode.
+* Download and install VirtualBox on your host https://www.virtualbox.org/wiki/Downloads
+* Download and install Vagrant: https://www.vagrantup.com/downloads 
+* Open powershell 
```
No real need to insist on powershell.

gmoraita replied:
Agreed

hugh-emerson reviewed (approved):
Good work! I've left some small suggestions but merge whenever you're ready 👍

gmoraita committed:
Merge branch 'master' into module_4

gmoraita committed:
Incorporated all suggestions from the review

- Updated README instructions
- Updated Vagrantfile and fixed also the executable of gunicorn to take env params
